### PR TITLE
C3 linearization/ Multiple Inheritance Risk

### PR DESCRIPTION
**Pull request by djphillyg**
Added Inheritance Risks from #130 , and went through the example in depth

### CODE REVIEW

File: docs/recommendations.md [hunk 1]
```diff
@@ -408,6 +408,47 @@ modifier auction_complete {
 ```
 `block.number` and *[average block time](https://etherscan.io/chart/blocktime)* can be used to estimate time as well, but this is not future proof as block times may change (such as [fork reorganisations](https://blog.ethereum.org/2015/08/08/chain-reorganisation-depth-expectations/) and the [difficulty bomb](https://github.com/ethereum/EIPs/issues/649)). In a sale spanning days, the 12-minute rule allows one to construct a more reliable estimate of time. 
 
+## C3 Linearization in [Polymorphic Contracts](https://pdaian.com/blog/solidity-anti-patterns-fun-with-inheritance-dag-abuse/)
+
+Solidity is a language that supports polymorphic inheritance, and specifically uses [C3 Linearization](https://en.wikipedia.org/wiki/C3_linearization) to obtain the order of inheritance. 
+
+Consider this [example](https://github.com/Arachnid/uscc/blob/master/submissions-2017/philipdaian/MDTCrowdsale.sol):
```
Comment (maurelian) [inline docs/recommendations.md]: The openings of this and the following paragraph feel pretty redundant.

File: docs/recommendations.md [hunk 2]
```diff
@@ -408,6 +408,47 @@ modifier auction_complete {
 ```
 `block.number` and *[average block time](https://etherscan.io/chart/blocktime)* can be used to estimate time as well, but this is not future proof as block times may change (such as [fork reorganisations](https://blog.ethereum.org/2015/08/08/chain-reorganisation-depth-expectations/) and the [difficulty bomb](https://github.com/ethereum/EIPs/issues/649)). In a sale spanning days, the 12-minute rule allows one to construct a more reliable estimate of time. 
 
+## C3 Linearization in [Polymorphic Contracts](https://pdaian.com/blog/solidity-anti-patterns-fun-with-inheritance-dag-abuse/)
```
Comment (maurelian) [inline docs/recommendations.md]: Make the title an imperative statement like "Use caution with inheritance"

File: docs/recommendations.md [hunk 3]
```diff
@@ -408,6 +408,47 @@ modifier auction_complete {
...
+
+Solidity is a language that supports polymorphic inheritance, and specifically uses [C3 Linearization](https://en.wikipedia.org/wiki/C3_linearization) to obtain the order of inheritance. 
+
```
Comment (maurelian) [inline docs/recommendations.md]: Not everyone will read and understand the take home from the example, so please add a 1 sentence concrete recommendation stating the actual recommendation.

**Comment by mimoo** [comment]
Are you sure this is still an issue? I cannot compile such contracts on Remix.

**Comment by mimoo** [comment]
Are you sure this is still a problem? I get an issue trying to compile this very simple example:

screen shot 2018-03-13 at 4 54 28 pm

> Linearization of inheritance graph impossible

On the other hand I can do something like this:

screen shot 2018-03-13 at 4 55 29 pm

where functions over different inherited contracts override themselves. Not sure if this mistake has a name.

**Comment by maurelian** [comment]
> Linearization of inheritance graph impossible

That error means there's a circular reference in the inheritance graph. Given 

```
B
C is B
A is C, B (last inherited should be 'most derived')
```
try to linearize:

```
C <- B
A <- B <- C
```

We can't create a valid DAG, because C is dependent on B. In your first example, you should be able to appease the compiler just by changing it to `A is B, C`. 

The danger here of incorrectly overriding a function is described well in this post: https://pdaian.com/blog/solidity-anti-patterns-fun-with-inheritance-dag-abuse/

**Comment by mimoo** [comment]
Is there really an issue here though? The contract A is overriding C's initialization of B. But that's what you would expect no?

screen shot 2018-03-14 at 3 29 36 pm

**Comment by mimoo** [comment]
I guess this is more what the blog was talking about:

screen shot 2018-03-14 at 3 34 00 pm

**Comment by maurelian** [comment]
Yes, I think so. a = 5, after that right?

(I'd love to play with your code samples, in the future could you also share the text format in a gist? The github logo at the top of remix makes it easy. )

**Comment by mimoo** [comment]
yes! Look at the right of my screenshot. And here's the [gist](https://gist.github.com/anonymous/dae4bab5a1cb9c93bf6277cc9ddd7afe) :)

**Comment by mimoo** [comment]
(so btw, I think this kind of example is more straight forward and shorter to include than the current PR)

**Comment by maurelian** [comment]
@mimoo yes, agree. @djphillyg and I were discussing yesterday, and felt the code samples should tend towards the minimal LOC necessary to express the issue. For more extensive examples, it makes sense just to link outwards.

**Comment by djphillyg** [comment]
Great critique. I used @mimoo 's example and referred to the Solidity github project and phil's work on inheritance security.

**Comment by mimoo** [comment]
Awesome, what about 4 more lines to introduce function shadowing in the same example? see https://gist.github.com/anonymous/6d2616a3a0d4ad5aefe83b898c5b55d6

**Comment by djphillyg** [comment]
Enhanced example for show of function overshadowing

**Pull request merged**